### DISCUSSION TITLE
Any supports for neoconf ?

### DISCUSSION THREAD

**Discussion by TD-Sky** [mrcjkb/rustaceanvim#49] [Q&A] [answered]
I use `.neoconf.json` to configure rust-analyzer locally. Now I cannot setup LSP client via `nvim-lspconfig` as rustaceanvim will do it. So are there any other ways to pass the project local settings?

**Comment by mrcjkb** [comment]
Hey :wave: 

Which settings do you need to configure on a per-project basis?
I plan on adding support for loading rust-analyzer settings from JSON (see #51).

Neoconf requires a `setup` call, which I view as potentially problematic, because it means that users have to worry about initialization order. In the case of this plugin, it might be fine though, as the config is initialised lazily when opening a Rust file.

I'll create a feature request for now (https://github.com/mrcjkb/rustaceanvim/issues/53), and look into it when I have some more time.
PRs are also more than welcome :smile:

**Comment by TD-Sky** [comment]
Here's the example coming from the project RISC-V based system kernel.

```json5
// .neoconf.json
{
  "lspconfig": {
    "rust_analyzer": {
      "rust-analyzer.checkOnSave.command": "clippy",
      "rust-analyzer.checkOnSave.allTargets": false,
      "rust-analyzer.cargo.target": "riscv64gc-unknown-none-elf"
    }
  }
}
```

**Reply by mrcjkb** [reply, marked as answer]
Those are rust-analyzer settings, so it can be solved with #51 (which I've just implemented and released).

See the [section in the readme](https://github.com/mrcjkb/rustaceanvim#how-to-dynamically-load-different-rust-analyzer-settings-per-project) for details.

The JSON would have to look like this:

```json
{
  "rust-analyzer": {
    "checkOnSave.command": "clippy",
    "checkOnSave.allTargets": false,
    "cargo.target": "riscv64gc-unknown-none-elf"
}
```

If you want to reuse your neoconf settings, you can use [this function](https://github.com/mrcjkb/rustaceanvim/blob/e3e0dc2ce1ba669ca8d8f367d76193829fe22192/lua/rustaceanvim/config/server.lua#L27) as a reference to load them.

~I do plan to add support for neoconf, but unfortunately I don't have time for that right now (probably not before December).~

**Reply by TD-Sky** [reply]
Really helpful ! Thanks for your work.

**Comment by zjp-CN** [comment]
> The JSON would have to look like this:
> 
> ```json
> {
>   "rust-analyzer": {
>     "checkOnSave.command": "clippy",
>     "checkOnSave.allTargets": false,
>     "cargo.target": "riscv64gc-unknown-none-elf"
> }
> ```

To inform anyone reading the solution when you migrate the setting from neoconf, you can copy and paste the content in `rust_analyzer` instead of making the above form. i.e.

```diff
// .neoconf.json
{
-  "lspconfig": {
-    "rust_analyzer": {
      "rust-analyzer.checkOnSave.command": "clippy",
      "rust-analyzer.checkOnSave.allTargets": false,
      "rust-analyzer.cargo.target": "riscv64gc-unknown-none-elf"
-    }
-  }
}
```

to obtain:

```json
// rust-analyzer.json under project root
{
      "rust-analyzer.checkOnSave.command": "clippy",
      "rust-analyzer.checkOnSave.allTargets": false,
      "rust-analyzer.cargo.target": "riscv64gc-unknown-none-elf"
}
```

then use `:RustAnalyzer reloadSettings` to reload the config without restarting.
